### DISCUSSION TITLE
🥊 Hole punching with DCUtR not working

### DISCUSSION THREAD

**Discussion by mkermani144** [libp2p/js-libp2p#2388] [Q&A] [answered]
Greetings everyone! 👋

I've been diving into the world of enabling hole punching with js-libp2p lately, dedicating a substantial amount of time to it, but alas, no breakthroughs. After some contemplation, I decided to initiate a [discussion](https://discuss.libp2p.io/t/hole-punching-in-dockerized-service-using-js-libp2p/2201/4) over on the libp2p forum, and someone suggested seeking assistance here.

In a nutshell, I experimented with the `rust-libp2p` hole punching example across three different servers – one with a public IP (for relay) and the others within a private network, behind NAT. Following a few tweaks, I successfully executed a hole punching scenario over TCP, both in non-Dockerized and Dockerized environments. While it didn't achieve a 100% success rate (likely due to using TCP), it hovered around 60%.

Returning to my own JS version of the scenario using js-libp2p, I encountered a roadblock. Despite numerous modifications, successful hole punching remained elusive. I've documented my code and challenges in this [GitHub repository](https://github.com/mkermani144/js-libp2p-hole-punching).

Any thoughts or ideas?

@sukunrt, I was directed here from the libp2p forum as someone who might have insights. Would greatly appreciate any assistance or guidance!

**Comment by sukunrt** [comment]
Can you check if your Private node knows its public IP? 
If you know it upfront you can try providing it via an env variable and see if things work.

**Reply by mkermani144** [reply]
Yes, I know the public IP of private nodes, but how does that help? Doesn't identify exactly do the same thing? In addition, there may be many scenarios that nodes don't know their public IPs.

**Reply by sukunrt** [reply]
You need some address discovery mechanism. Identify is what we use in go-libp2p and probably js-libp2p too. But at least in go-libp2p we need a few(4) identify replies before the node believes the address to be its external address. In your repro, are you connecting to 4 or more peers? 

If we can provide this address externally we can zero down on the problem. Whether it is one of address discovery or one where we are unable to hole punch even with correct addresses.

**Reply by mkermani144** [reply]
Yes, the 4 reply limit is what I found in the source code and the spec, and thought that may cause a problem. But when I looked at the rust examples, it worked without any issues and without providing addresses explicitly, and even without using AutoNAT at all. 

By the way, regarding your solution, I still can't find a way in `js-libp2p` to provide that external address explicitly, unless you mean dialing that address directly, which I don't think is the case as it negates DCUtR protocol completely.

**Reply by achingbrain** [reply]
You can tell a node it's public address with the `announce` config option:

```TypeScript
const node = await createLibp2p({
  addresses: {
    listen: [
      // this address will start off untrusted and requires a few peers to tell us
      // it's publicly dialable before we announce it to the network
      '/ip4/0.0.0.0/....'
    ],
    announce: [
      // this address will be trusted as public
      '/ip4/x.x.x.x/...',

      // so will this one
      '/dns4/example.com/...'
    ]
  }
})
```

**Reply by mkermani144** [reply]
@achingbrain Thanks, I will try Sukunrt's solution with this method and see what happens.

**Reply by mkermani144** [reply]
@sukunrt @achingbrain 
I tried this solution. It basically works, but it seems when I add `announce` addresses, because the `identify` includes public IP addresses for the listener (A), DCUtR is done in unilateral mode. I'm not sure how much this method is scalable, but nonetheless, there are ways to find current node public IP before starting the node.

**Reply by mkermani144** [reply]
Just for future reference, I tested the solutions provided and it works. It's not a full DCUtR scenario, but it does the job nonetheless. I tried to find my public ip with the `identify` protocol and succeeded, but because there seems to be no way of altering announce addresses after libp2p node is run, there is a need for other mechanisms such as `public-ip` package before running the node.

**Reply by sukunrt** [reply]
In most cases you'd connect to many other nodes, so the address discovery should work. In this case you're just connecting to 1 node so the confidence in the address discovered over identify is low.

**Reply by mkermani144** [reply]
I think you mean through `autonat` service, don't you? Because I see no address confirmation mechanism for observed addresses in `identify` protocol (at least in `js-libp2p`).

**Reply by mkermani144** [reply]
And one more concern is that, `js-libp2p` _seems_ not to re-use the `listen` port for creating new connections, which cause others to identify us with a random port different from our `listen` port.

**Reply by achingbrain** [reply]
> I see no address confirmation mechanism for observed addresses in identify protocol 

It's here: https://github.com/libp2p/js-libp2p/blob/main/packages/protocol-identify/src/identify.ts#L313

We store the observed address with no confidence: https://github.com/libp2p/js-libp2p/blob/main/packages/libp2p/src/address-manager/index.ts#L168-L170

Then later on AutoNat gets the observed addresses: https://github.com/libp2p/js-libp2p/blob/main/packages/protocol-autonat/src/autonat.ts#L343-L348

...and confirms the observed address(es) if enough peers can dial them: https://github.com/libp2p/js-libp2p/blob/main/packages/protocol-autonat/src/autonat.ts#L502

**Reply by mkermani144** [reply]
Yes, I looked at the exact codes you sent beforehand. I meant `identify` doesn't confirm an observed address on its own, it's the `autonat` that does so.

**Comment by achingbrain** [comment, marked as answer]
Thanks for digging into this and trying to make it more reliable, the research is very valuable.

We've not put a lot of effort in to hole punching in js-libp2p because, timing issues in JS aside, it really needs QUIC for to be reliable and support is still missing in Node.js.

The current DCUtR protocol implementation is mainly there to enable the unilateral connection attempt part of the spec whereby nodeA dials nodeB via circuit relay, but nodeA has a publicly routable address that nodeB can dial directly so it does that.

**Reply by mkermani144** [reply]
I fully understand. Great job by the way.

I tested the unilateral hole punching weeks ago and it works great, and just need some kind of conclusion on if we can rely on non-unilateral hole punching of js-libp2p or not. It's a crucial part of our network, as it helps a lot in reducing the load of relays.
